src/lib.rs

```diff
@@ -104,21 +131,46 @@ pub fn readline(prompt: &str) -> Option<String> {
     let cprmt = CString::new(prompt).unwrap().as_ptr();
     unsafe {
         let ret = ext_readline::readline(cprmt);
-        if ret.is_null() {  // user pressed Ctrl-D
-            None
-        }
-        else {
-            let slice = CStr::from_ptr(ret);
-            let bytes = slice.to_bytes();
+        c_str_to_string(ret)
+    }
+}
 
-            // the return from readline needs to be explicitly freed
-            // so clone the input first
-            let line = String::from_utf8_lossy(bytes).into_owned().clone();
+static mut _lhandler: Option<fn(Option<String>)> = None;
 
-            libc::free(ret as *mut libc::c_void);
+extern fn coerced_callback(ret: *const c_char) {
+    let line = c_str_to_string(ret);
 
-            Some(line)
-        }
+    unsafe {
+        _lhandler.unwrap()(line);
+    }
+}
+
+/// Install a new input handler.
+///
+/// This function must be called before `rl_callback_read_char()` or risk
+/// panicking.
+pub fn rl_callback_handler_install(prompt: &str, lhandler: fn(Option<String>)) {
+    let cprmt = CString::new(prompt).unwrap().as_ptr();
+    unsafe { _lhandler = Some(lhandler); }
```

Comment:
hmmmm. Not sure how I feel about this. ``lhandler`` is left as a ``Some`` value but it won't be useful to the next caller. what if ``coerced_callback`` was used as a curried/partial function instead?